xds/internal/balancer/cdsbalancer/cluster_handler_test.go

```diff
@@ -18,1078 +18,650 @@ package cdsbalancer
-
-	shouldNotHappenCtx, shouldNotHappenCtxCancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
-	defer shouldNotHappenCtxCancel()
-
-	select {
-	case <-ch.updateChannel:
-		t.Fatal("Cluster Handler wrote an update to updateChannel when it shouldn't have, as each node in the full cluster tree has not yet received an update")
-	case <-shouldNotHappenCtx.Done():
-	}
-
-	// Switch the cluster back to an EDS Cluster. This should cause the two
-	// children to be deleted.
-	fakeClient.InvokeWatchClusterCallback(xdsresource.ClusterUpdate{
-		ClusterType: xdsresource.ClusterTypeEDS,
-		ClusterName: edsService2,
-	}, nil)
-
-	// Should delete the two children (no guarantee of ordering deleted, which
-	// is ok), then successfully write an update to the update buffer as the
-	// full cluster tree has received updates.
-	clusterNameDeleted, err := fakeClient.WaitForCancelClusterWatch(ctx)
-	if err != nil {
-		t.Fatalf("xdsClient.CancelCDS failed with error: %v", err)
-	}
-	// No guarantee of ordering, so one of the children should be deleted first.
-	if clusterNameDeleted != edsService {
-		if clusterNameDeleted != logicalDNSService {
-			t.Fatalf("xdsClient.CancelCDS called for cluster %v, want either: %v or: %v", clusterNameDeleted, edsService, logicalDNSService)
-		}
-	}
-	// Then the other child should be deleted.
-	clusterNameDeleted, err = fakeClient.WaitForCancelClusterWatch(ctx)
-	if err != nil {
-		t.Fatalf("xdsClient.CancelCDS failed with error: %v", err)
-	}
-	if clusterNameDeleted != edsService {
-		if clusterNameDeleted != logicalDNSService {
-			t.Fatalf("xdsClient.CancelCDS called for cluster %v, want either: %v or: %v", clusterNameDeleted, edsService, logicalDNSService)
-		}
-	}
-
-	// After cancelling a watch for the second child cluster, there should be no
-	// more watches cancelled on the xds client.
-	shouldNotHappenCtx, shouldNotHappenCtxCancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
-	defer shouldNotHappenCtxCancel()
-	gotCluster, err = fakeClient.WaitForCancelClusterWatch(shouldNotHappenCtx)
-	if err == nil {
-		t.Fatalf("xdsClient.WatchCDS called for cluster: %v, no more watches should be cancelled.", gotCluster)
-	}
-
-	// Then an update should successfully be written to the update buffer.
-	select {
-	case chu := <-ch.updateChannel:
-		if diff := cmp.Diff(chu.updates, []xdsresource.ClusterUpdate{{
-			ClusterType: xdsresource.ClusterTypeEDS,
-			ClusterName: edsService2,
-		}}); diff != "" {
-			t.Fatalf("got unexpected cluster update, diff (-got, +want): %v", diff)
-		}
-	case <-ctx.Done():
-		t.Fatal("Timed out waiting for update from update channel.")
+// Tests the scenario where an aggregate cluster exceeds the maximum depth,
+// which is 16. Verfies that the channel moves to TRANSIENT_FAILURE, and the
+// error is propagated to RPC callers. The test then modifies the graph to no
+// longer exceed maximum depth, and verifies that an RPC can be made
+// successfully.
```

Comment:
s/the graph/the cluster, and mention it's right on the boundary of recursion after you update it (i.e. the maximum depth possible for the graph).